services/vision/mlvision/detector.go

```diff
@@ -0,0 +1,170 @@
+	}
+
+	// Set up input type, height, width, and labels
+	var inHeight, inWidth uint
+	inType := md.Inputs[0].DataType
+	labels := getLabelsFromMetadata(md)
+	boxOrder, err := getBoxOrderFromMetadata(md)
+	if err != nil || len(boxOrder) < 4 {
+		boxOrder = []int{1, 0, 3, 2}
+	}
+	if shape := md.Inputs[0].Shape; getIndex(shape, 3) == 1 {
+		inHeight, inWidth = uint(shape[2]), uint(shape[3])
+	} else {
+		inHeight, inWidth = uint(shape[1]), uint(shape[2])
+	}
+
+	return func(ctx context.Context, img image.Image) ([]objectdetection.Detection, error) {
+		origW, origH := img.Bounds().Dx(), img.Bounds().Dy()
+		resized := resize.Resize(inWidth, inHeight, img, resize.Bilinear)
+		inMap := make(map[string]interface{})
+		switch inType {
+		case UInt8:
+			inMap["image"] = rimage.ImageToUInt8Buffer(resized)
+		case Float32:
+			inMap["image"] = rimage.ImageToFloatBuffer(resized)
+		default:
+			return nil, errors.New("invalid input type. try uint8 or float32")
+		}
+		outMap, err := mlm.Infer(ctx, inMap)
+		if err != nil {
+			return nil, err
+		}
+
+		locations := unpack(outMap, "location", md)
+		categories := unpack(outMap, "category", md)
+		scores := unpack(outMap, "score", md)
+
+		// Now reshape outMap into Detections
+		detections := make([]objectdetection.Detection, 0, len(categories))
+		for i := 0; i < len(scores); i++ {
+			xmin, ymin, xmax, ymax := utils.Clamp(locations[4*i+getIndex(boxOrder, 0)], 0, 1)*float64(origW),
+				utils.Clamp(locations[4*i+getIndex(boxOrder, 1)], 0, 1)*float64(origH),
+				utils.Clamp(locations[4*i+getIndex(boxOrder, 2)], 0, 1)*float64(origW),
+				utils.Clamp(locations[4*i+getIndex(boxOrder, 3)], 0, 1)*float64(origH)
+			rect := image.Rect(int(xmin), int(ymin), int(xmax), int(ymax))
+			labelNum := int(categories[i])
+
+			if labels != nil {
+				detections = append(detections, objectdetection.NewDetection(rect, scores[i], labels[labelNum]))
+			} else {
+				detections = append(detections, objectdetection.NewDetection(rect, scores[i], strconv.Itoa(labelNum)))
+			}
+		}
+		return detections, nil
+	}, nil
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string, md mlmodel.MLMetadata) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch getTensorTypeFromName(name, md) {
+	case UInt8:
+		temp := me.([]uint8)
+		for _, t := range temp {
+			out = append(out, float64(t))
+		}
+	case Float32:
+		temp := me.([]float32)
+		for _, p := range temp {
+			out = append(out, float64(p))
+		}
+	default:
+		return nil
+	}
+	return out
+}
+
+func getTensorTypeFromName(name string, md mlmodel.MLMetadata) string {
+	for _, o := range md.Outputs {
+		if strings.Contains(strings.ToLower(o.Name), strings.ToLower(name)) {
+			return o.DataType
+		}
+	}
+	return ""
+}
+
+// getLabelsFromMetadata returns a slice of strings--the intended labels
+func getLabelsFromMetadata(md mlmodel.MLMetadata) []string {
```

Comment:
@kharijarrett  Same with all of these helper functions - put them in the ml_model.go file, rather than in the detector file

Reply:
Done.